Fix log writing failed. can't be called from trap context (thread version)

sonots opened the pull request:
Fix https://github.com/resque/resque/issues/1493

sonots commented:
CI failed, but it is not related with this commit (failed with ruby 1.9.3 and 1.8.7)

fw42 commented:
Hmmm.. I'm not a fan of how this will make your application multi-threaded in possibly unexpected ways. For example, at Shopify, our Logger is actually logging to Kafka, and with this PR, that Kafka logger would run in a seperate thread. Now I have to think about if the logging code is thread-safe, which I wouldn't expect having to think about in an otherwise single-threaded application.

@dylanahsmith, how do you feel about this?

dylanahsmith commented on lib/resque/signal_handler.rb:
```diff
@@ -0,0 +1,38 @@
+    end
+
+    @handlers[sig] = block
+    Kernel.trap(sig) { @self_write.puts(sig) }
```
Isn't it the logging not the signal handling that needs to be done in another thread?

dylanahsmith commented:
If you want a separate thread to do the logging, then couldn't that be accomplished with a custom logger that writes the log entries to a pipe and has its own separate thread for doing the actual logging?

sonots commented:
Basically, mutex#lock can't be called from trap context. ref. https://bugs.ruby-lang.org/issues/7917
Problem is not only for logger.
Pushing signal into an array or pipe in trap context, and hanling them in somewhere else is the common way handling signal as unicorn, sidekiq, fluentd does.

sonots commented:
* sidekiq https://github.com/mperham/sidekiq/blob/b6de0bae6b86d5abae25852141768d8ecc8ddedf/lib/sidekiq/cli.rb#L53
* unicorn https://github.com/defunkt/unicorn/blob/master/lib/unicorn/http_server.rb#L120

sonots commented:
We basically have to avoid heavy work in trap context because we can not trap a new signal during processing something inside trap context.
We therefore should only put signals into an array or pipe in trap context, and then treat signals somewhere else sequentially not to lose new signals.

sonots commented:
I came up with a way not to use thread > https://github.com/resque/resque/compare/master...sonots:fix_signal_trap2?expand=1

However, in this way, the behavior of signal handler for a forked child must have been changed if TERM_CHILD environment variable is not given.

Before:

* TERM: Set `@shutdown = true`, but actually affects nothing
* INT: Set `@shutdown = true`, but actually affects nothing
* QUIT: Set `@shutdown = true`, but actually affects nothing
* USR1: Do nothing
* USR2: run_hook :before_pause and set `@paused = true`
* CONT: Set `@paused = false` and run_hook :after_pause

After:

* All signals are reverted to DEFAULT

Which do you prefer?

BTW: Rather, I did not get why resque was not calling unregister_signal_handlers for a forked child always. I felt as `Before` behavior was wierd because it disabled TERM or INT signals.

fw42 commented:
The problem I see with your other approach is that that code (where the `sleep` was) will only get executed if the queue is empty (`work_one_job` returns `nil`/`false`). So if there is sufficiently many jobs being pushed into the queue, you will never get to the signal handling part.

Can you open a PR for that so that we can discuss the code?

sonots commented:
Okay, sent another PR https://github.com/resque/resque/pull/1502

sonots commented:
I added Thread.stop to stop main thread on receiving signal once, not to require handling a signal to be thread-safe anymore > https://github.com/resque/resque/pull/1499/commits/8fff907d49b73602211ef5efbfdd013053848889

fw42 commented on lib/resque/worker.rb:
```diff
@@ -861,7 +862,11 @@ def perform_with_fork(job, &block)
+          if term_child
+            unregister_signal_handlers
+          else
+            @signal_handler.reopen
```
why is this necessary?

sonots replied:
It is because

1. fork does not copy thread
2. IO.pipe is shared with the parent process

which results in that receiving a signal in child process makes an effect to its parent process.
`reopen` newly creates signal handler thread and pipe for the child process.

fw42 replied:
Ok makes sense!

fw42 replied:
are any tests failing if you remove this line here? If not, can you add a new test?

sonots replied:
It seems difficult to create tests for handling signals, but let me try it later.

sonots replied:
Added https://github.com/resque/resque/pull/1499/commits/885d51fb86651be61bdbb4485bbe7b27df405331, but sleep 0.1 * 10 times would be feasible...

fw42 commented on test/child_killing_test.rb:
```diff
@@ -92,11 +98,26 @@ def assert_child_not_running(child_pid)
+
+    Process.kill('TERM', worker_pid)
+    Process.waitpid(worker_pid)
+    result = Resque.redis.lpop('sigterm-test:result')
```
what does this line do? isn't there an assertion missing here?

sonots replied:
it seems that LongRunningJob does `rpush`, this line is just for `lpop` it. Is it not needed?

fw42 commented on test/child_killing_test.rb:
```diff
@@ -92,11 +98,26 @@ def assert_child_not_running(child_pid)
     end
   end
+
+  it 'sending signal to a forked child should not make a effect to its parent' do
```
just to double-check: This test fails without the line above?

sonots replied:
yes

fw42 commented:
I like this approach. @dylanahsmith, @steveklabnik, wdyt?

sonots commented:
> I added Thread.stop to stop main thread on receiving signal once, not to require handling a signal to be thread-safe anymore > 8fff907

This way was not good. Main thread would already be a stop state before calling 
Thread.stop in some reasons. Main thread stucks if Thread.main.wakeup is called before
Thread.stop. (This situation happened to me). Reverted.

dylanahsmith commented:
I still dislike this approach.  There seems to be an assumption that we need to move all the signal handling logic outside of the signal handler, but 

> Basically, mutex#lock can't be called from trap context. ref. https://bugs.ruby-lang.org/issues/7917
> Problem is not only for logger.

Where else do we use mutex#lock other than in the implementation of the logger?  The signal handlers are not actually that complicated.  Some of those mostly just set an instance variable, which seems like overkill to do by piping to another thread.

The other possible concern might be the before_pause and after_pause hooks, but moving that to a separate thread could break applications using that hook.  If the application wants to avoid this problem, then they could also avoid logging or just delay logging.

What if we just append those messages to an array, which is safe to do in a signal handler, then pop them off and write them in the main thread.  For instance we could do that when methods like Resque::Worker#shutdown? or Resque::Worker#paused? are called to check those instance variables so that we log before the signal actually has an effect.

chrisccerami commented:
Is this PR still in progress?

dchersey commented:
Sorry, did this PR get merged?
